drud/ddev
# Make supervisord exit when key processes die, fixes #1137

## The Problem/Issue/Bug:

In the web container, supervisord can happily work along even though php-fpm or the webserver has died or is not runnable.

## How this PR Solves The Problem:

Use technique from https://blog.zhaw.ch/icclab/process-management-in-docker-containers/ to add an event listener, so supervisord and thus the container will die in cases like this.


## Manual Testing Instructions:

* Start a project
* `ddev ssh`
* Do things like `killall nginx` or `killall php-fpm` or introduce an error into /etc/nginx/nginx.conf and killall -1 nginx, or kill -1 1 (to reload supervisord itself). 

Those should result in nginx or php-fpm or apache croaking, and the container exiting.

## Automated Testing Overview:


## Related Issue Link(s):

OP #1137 

## Release/Deployment notes:

This does in fact change the behavior of the web container. Previously, if a process died (like apache or nginx) the container would continue on. It would probably become "unhealthy" but would stay up. Here instead if one of those croaks you get a container that stops.

One interesting thing about this implementation is there is just *one* event listener, which is causing supervisord to exit if any process dies. It would be easy to implement separate listeners that had separate behavior. For example, we could then ignore the failure of mailhog.
